java/org/contikios/cooja/Simulation.java

```diff
@@ -353,36 +400,44 @@ public void stopSimulation() {
    * @param rv Return value from startSimulation, should be null unless > 0
    */
   public void stopSimulation(boolean block, Integer rv) {
-    if (!isRunning()) {
+    if (!isRunning() || isShutdown) {
       return;
     }
     assert rv == null || rv > 0 : "Pass in rv = null or rv > 0";
     if (rv != null) {
       returnValue = rv;
     }
-    stopSimulation = true;
 
-    if (block) {
-      if (Thread.currentThread() == simulationThread) {
-        return;
-      }
+    commandQueue.add(STOP);
 
-      /* Wait until simulation stops */
-      try {
-        Thread simThread = simulationThread;
-        if (simThread != null) {
-          simThread.join(100);
-        }
-      } catch (InterruptedException e) {
-      }
+    if (block) {
+      waitFor(false, 250);
     }
-    cooja.updateProgress(true);
   }
 
-  public void stopScriptEngines() {
-    // FIXME: this should be at the end of run(), but that requires the sim thread to be persistent.
-    for (var engine : scriptEngines) {
-      engine.deactivateScript();
+  private void waitFor(boolean isRunning, long maxTime) {
+    if (Thread.currentThread() == simulationThread) {
+      return;
+    }
+    try {
+      long startTime = System.currentTimeMillis();
+      synchronized (lock) {
+        if (maxTime == 0) {
+          while (this.isRunning != isRunning && !isShutdown) {
+            lock.wait();
+          }
+        } else {
+          while (this.isRunning != isRunning && !isShutdown) {
```

Comment:
These two loops are almost identical, would be nice to only have a single loop.

Reply:
I will combine them. Thanks.